lib/IRGen/IRABIDetailsProvider.cpp

```diff
@@ -122,6 +125,32 @@ class IRABIDetailsProviderImpl {
     return {returnTy, {paramTy}};
   }
 
+  llvm::SmallVector<IRABIDetailsProvider::ABIAdditionalParam, 1>
+  getFunctionABIAdditionalParams(AbstractFunctionDecl *afd) {
+    llvm::SmallVector<IRABIDetailsProvider::ABIAdditionalParam, 1> params;
+    Mangle::ASTMangler mangler;
+
+    auto mangledName = mangler.mangleAnyDecl(afd, /*prefix=*/true);
+    for (auto &f : silMod->getFunctions()) {
```

Comment:
Sorry, I know I suggested this approach of lowering to sil and finding the function initially, but it’s actually not the approach we should pursue. 
We just want to hook into SILGenModule, and it's expose its API we can then use from IRABIDetailsProvider.

For example:
// SomeHeader.h
SILFunction *getFunction(SILDeclRef ref);

// SomeFile.cpp (inside lib/SIL)
SILFunction *getFunction(SILDeclRef ref) {
  SILGenModule SGM(...);
  return SGM.getFunction(ref, NotForDef);
}

Reply:
I used the `SILFunction.h` and `SILFunction.cpp` to place this function and it seems to work for the general case. However, this implementation sometimes crashes the compiler on the `function-availability.swift` test, but I couldn't reproduce the bug outside the lit test framework...

Reply:
This is the output error: https://gist.github.com/Robertorosmaninho/276f9cf9888ce74e2e1a08ce41c3439d